geomstats/learning/natural_gradient_descent.py

```diff
@@ -0,0 +1,120 @@
+"""Natural gradient descent using Amari's anaytic solution.
+
+References
+----------
+.. [AS1985] Amari S., Karakida R., Oizumi M. (2018)
+    Fisher Information and Natural Gradient Learning of Random Deep Networks.
+"""
+
+import math
+
+import numpy as np
+import torch
+from torch.optim import Optimizer
+
+import geomstats.backend as gs
+
+
+class NaturalGradientDescent(Optimizer):
+    """Natural Gradient Descent.
+
+    Uses Amari's analytic solution for the inverse Fisher information matrix.
+
+    Parameters
+    ----------
+    params : array-like, shape=[3, dim]
+        The current layer of the ResNet, consisting of a vector of nodes,
+        a vector of weights, and a bias scalar
+    lr : float
+        The learning rate to update the weights
+    wd : float
+        Penalization on weights to prevent overfitting
+    bias : Boolean
+        Whether or not to include bias term in training
+
+    References
+    ----------
+    .. [AS1985] Amari S., Karakida R., Oizumi M. (2018)
+        Fisher Information and Natural Gradient Learning of Random Deep Networks.
+    """
+
+    def __init__(self, params, lr=0.05, wd=0.0, bias=True):
+        defaults = dict(lr=lr, wd=wd, bias=bias)
+        self.params = params
+
+        super().__init__(params, defaults)
+
+    def step(self):
```

Comment:
I think we should probably split this into smaller functions, where it is easier to understand what are the inputs and outputs: e.g. it seems we collect weight layers for each param group, maybe a `_collect_weight_layers` method helps on code readability. Also, creating a `_step_param_group` may help avoiding so much nesting (which, as I comment below, seem to be bug prone).